Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp

```diff
@@ -28,26 +28,154 @@
 
 #include "GPUAdapter.h"
 #include "GPUCanvasConfiguration.h"
+#include "RenderBox.h"
+#include <wtf/IsoMallocInlines.h>
+
+#if PLATFORM(COCOA)
+#include "WebGPUSurfaceCocoa.h"
+#endif
 
 namespace WebCore {
 
+static IntSize getCanvasSizeAsIntSize(const GPUCanvasContext::CanvasType& canvas)
+{
+    return WTF::switchOn(canvas, [](const RefPtr<HTMLCanvasElement>& htmlCanvas) -> IntSize {
+        return { static_cast<int>(htmlCanvas->width()), static_cast<int>(htmlCanvas->height()) };
+    }
+#if ENABLE(OFFSCREEN_CANVAS)
+    , [](const RefPtr<OffscreenCanvas>& offscreenCanvas) -> IntSize {
+        return { static_cast<int>(offscreenCanvas->width()), static_cast<int>(offscreenCanvas->height()) };
+    }
+#endif
+    );
+}
+
+static bool platformSupportsWebGPUSurface()
+{
+#if PLATFORM(COCOA)
+    return true;
+#else
+    return false;
+#endif
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(GPUCanvasContext);
+
+std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
+{
+    if (!platformSupportsWebGPUSurface())
+        return nullptr;
+
+    auto context = std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
+    context->suspendIfNeeded();
+    return context;
+}
+
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+{
+    ASSERT(platformSupportsWebGPUSurface());
+}
+
+void GPUCanvasContext::reshape(int width, int height)
+{
+    if (!m_currentConfiguration)
+        return;
+
+#if PLATFORM(COCOA)
+    m_surface = WebGPUSurfaceCocoa::create({ width, height }, m_currentConfiguration->convertToBacking());
```

Comment:
Maybe this is a silly question, but why does the abstraction layer need to be at the WebGPU layer? Why not just have WebGPUSurface (not WebGPUSurfaceCocoa) and have the objects inside it be platform-specific? I believe this is the design we use for compositing IIRC.

Reply:
The reasoning from Kiet was to allow non-Apple platforms have a platform specific implementation of WebGPUSurface that uses something other than IOSurfaces.

Do you think it is worth keeping that abstraction or should the platform specific objects be placed inside WebGPUSurface allowing WebGPUSurfaceCocoa to be removed?

Reply:
I think it's better to keep all the `#PLATFORM(COCOA)` yuckiness contained inside a single file, rather than sprayed around to all the callers who need to use surfaces.